### PR TITLE
BUG: Extend current in calcSeries to quad4

### LINKED ISSUE #110: "Resolution" of the I-V plots made with plotSys() changes with cell temperature
If I set the cell temperatures high, then the "resolution" of the I-V and Power plots made with plotSys() decreases:
systems_with_different_temps
These 3 models have a uniform 1 suns of irradiance on them. The only difference is the cell temperatures indicated with colors.

### PR DESCRIPTION
**Pull request by mikofski**
- fixes #110 - bad resolution for IV curve at high temperature
- change Iforward range to be from 0 to meanIsc, instead of from Imin
which was TOO SMALL, and use pvconst.Imod_pts which has closer spacing
near zero than pvconst.pts
- add quad4 range from Imin to zero, where Imin is a negative current in
the 4th quadrant, using pvconst.Imod_negpts which goes from 1 to 0,
with closer spacing near zero
- concatenate Iquad4 before Iforward, and increase Vtot to same size
- also add new pvconst.Imod_pts_sq which has even closer spacing near
zero to use with Ireverse, so there's more than 1 point before the
breakdown voltage!
- add pvstr.Voc_mod which is a list of the sum of Voc for all cells in
each module
- and add pvsys.Voc_str which is a list of the sum of the Voc for all
modules in each string
- then use pvsys.Voc_str.max() instead of pvsys.Vstring.max() in
calcParallel so there's not too many points in the 4th quadrant pull
thing the curve way off, and messing up the resolution in the forward
bias region (1st quadrant)
- add a test in test_pvconstants to test the minimum current close to
max voc by comparing the IVP curve of pvsys with data from gh110.dat
which contains the correct resolution for a curve with npts=101, the
default at Tcell=288[K]
- also update test_pvcell.test_calc_series() with new data, but also
compare the old data by interpolating with the new - we can delete the
old data eventually right?
- still a few more bugs, tests to fix, so this is a WIP

### CODE REVIEW

**Comment by mikofski** [comment]
OK, fixed partial cross ties, but now total cross ties is broken

**Comment by mikofski** [comment]
@adambgnr and @chetan201 I think this is ready for review.

@adambgnr please bear in mind that when I started pvmismatch like 7 years ago, I was just learning Python so there's a lot of really bad code in here that should probably be fixed up, but I don't think this PR covers that.

This PR just fixes #110 the high temperature system, string, and module problem. The bug was caused by #73 a fix for #36 which solved the high temperature problem for voltage in the cell. This fix is the same approach, but at the module, string, and system level, and splits the IV curve into 3 regions instead of 2, when interpolating either voltage or current:
* quadrant 4
* forward bias
* reverse bias

Let me know if you have any questions, and please open separate issues if you want to address improving the code for readability or performance. Thanks!

File: pvmismatch/pvmismatch_lib/pvsystem.py [hunk 1]
```diff
@@ -81,13 +81,17 @@ def Istring(self):
...
+        return np.asarray([pvstr.Voc_mod.sum() for pvstr in self.pvstrs])
+
     def calcSystem(self):
         """
         Calculate system I-V curves.
         Returns (Isys, Vsys, Psys) : tuple of numpy.ndarray of float
         """
         Isys, Vsys = self.pvconst.calcParallel(
-            self.Istring, self.Vstring, self.Vstring.max(), self.Vstring.min()
+            self.Istring, self.Vstring, self.Voc_str.max(), self.Vstring.min()
```
Comment (mikofski) [inline pvmismatch/pvmismatch_lib/pvsystem.py]: What do you think about this, that the system is bounded by the largest Voc, and not necessarily into the 4th quadrant? I figure why would you ever need negative current at the system level? No inverter would ever even get close to this right?

Comment (adambgnr) [reply pvmismatch/pvmismatch_lib/pvsystem.py]: Makes sense to me. So far it never occurred to me to instigate negative current on a system level. Also blocking diodes or a charge controller wouldn't really allow this anyway.

**Comment by mikofski** [comment]
@chetan201 ready for review. @adambgnr feel free to add your comments, as you're the one who caught the bug.

File: pvmismatch/pvmismatch_lib/pvconstants.py [hunk 2]
```diff
@@ -111,6 +111,8 @@ def npts(self, npts):
         # so that tight spacing is around MPP and RBD
```
Comment (adambgnr) [inline pvmismatch/pvmismatch_lib/pvconstants.py]: I made a custom version of plotSys() and ran:
```
pvconst = pvconstants.PVconstants(npts=20) # I did this just to see the spacing better
pvsys = pvsystem.PVsystem(pvconst=pvconst)
f = pvsys.plotSys()
f.show()
```
spacing
Maybe I'm misunderstanding something, but I was wondering if this is what you wanted to achieve. Now MPP falls into the non-tightly spaced part of the second section of sample points and it gets more tightly spaced towards VOC. Shouldn't the second section be reversed?

**Comment by adambgnr** [comment]
> @chetan201 ready for review. @adambgnr feel free to add your comments, as you're the one who caught the bug.

Thanks a lot for making this fix @mikofski! Indeed now the plots (and the underlying results) are nice and smooth regardless of the temperature. While testing, I have found one thing that I don't understand, see in the comments at `npts()`.

File: pvmismatch/pvmismatch_lib/pvconstants.py (hunk 2 above)
Comment (mikofski) [reply pvmismatch/pvmismatch_lib/pvconstants.py]: Thanks. I see your point! But I'm ambivalent. When calculating current near Voc, very small changes in voltage result in big changes in current, and that region near Voc has a smaller voltage range than the Isc side of the mpp, but same `npts`, so it's already dense. So I'm not sure what's better. @chetan201? I think you might be right, but I'll ask a few others.

**Comment by mikofski** [comment]
Here's what the different spacings are supposed to look like:
Figure_1

So using `np.flip(Imod_negpts)` would reverse the spacing to create tighter spacing near the max power point, and less spacing as Vsys -> Voc.
```python
from pvmismatch import *
pvconst = pvconstants.PVconstants(npts=20)
pvsys = pvsystem.PVsystem(pvconst=pvconst)
f = pvsys.plotSys(fmt='+:')
f.show()
```
It looks like this:
Figure_2

I like it better. The benefit is that there are more points around the knee of the max power point which is kind of what we originally wanted so it's easier to determine the max power, and the only downside is that we may miss some curvature near Voc, but the user can increase `npts` if they want more detail there.

I also updated the `plotSys(fmt='')` to allow specification of plotting formats like `"+:"` used here, and I saved the new set of points in `PVconstants` as `Vmod_q4pts`. I noticed that `Imod_pts` isn't used anywhere anymore except to create `Imod_pts_sq` which is a really tight spacing to capture the reverse breakdown, which is very flat and fast for avalanche. Maybe there should be a ticket to remove this at some point, not sure if deprecation is required, since it should be private anyway. Also I added `sharex` and `tight_layout()` to the figures so they look nicer and zoom more easily.

File: pvmismatch/pvmismatch_lib/pvconstants.py (hunk 2 above)
Comment (mikofski) [reply pvmismatch/pvmismatch_lib/pvconstants.py]: fixed in 95644c2df334c8ee31588a217ac4875ea8162682

**Comment by adambgnr** [comment]
> It looks like this:
> Figure_2
> 
> I like it better. The benefit is that there are more points around the knee of the max power point which is kind of what we originally wanted so it's easier to determine the max power, and the only downside is that we may miss some curvature near Voc, but the user can increase `npts` if they want more detail there.
> 
> I also updated the `plotSys(fmt='')` to allow specification of plotting formats like `"+:"` used here

I like it too! Also I like the new `fmt='+:'` option for `plotSys()`! Much more informative than just the line plot.

**Comment by chetan201** [comment]
@mikofski @adambgnr 
Thanks for catching and fixing this. Since it changes the absolute value of the MPP, I would like to review a few cases before I approve. Thanks for your patience!

**Comment by mikofski** [comment]
@chetan201 (@PVMismatch/maintainer) gentle nudge :wink:
